### PR TITLE
Resolve bug that caused common variables to be renamed

### PR DESCRIPTION
**Pull request by keyboardDrummer**
The Dafny regression suite currently fails when upgrading to the latest Boogie: https://github.com/dafny-lang/dafny/pull/2046

### Changes

1. Correctly name variables after calling FullReset on an interactive or Batch prover.
1. Do not rename the variable "type" regardless of normalisation or randomness, since it is a hardcoded name that can be depended upon.

### Testing
- Add a test that verifies the first change

### CODE REVIEW

File: Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs [hunk 1]
```diff
@@ -95,8 +100,10 @@ public override Task BeginCheck(string descriptiveName, VCExpr vc, ErrorHandler
     public override void FullReset(VCExpressionGenerator generator)
     {
       this.gen = generator;
+      SendThisVC("(reset)");
```
[the remark this answers is not in the document]
Comment (atomb) [reply Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs:103]: Given that tests pass, though, I don't think this should hold up merging. We should maybe just investigate later.

**Review by atomb — APPROVED** [review]
The `(reset)` in the batch prover interface seems surprising to me, but I don't think it'll hurt, and we can investigate why it's needed later.

File: Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs:103]: > Given that the batch prover is started fresh for every query, it should never be necessary to reset it.

What makes you say it's a new Z3 process for every query?

[the remark this answers is not in the document]
Comment (atomb) [reply Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs:103]: That's one of the fundamental differences between `SMTLibBatchTheoremProver` and `SMTLibInteractiveTheoremProver` (and the primary reason that the former exists), which manifests here:

https://github.com/boogie-org/boogie/blob/678728eb6a0cbab2cddbf36a01e569d724e66e7f/Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs#L62-L64

[the remark this answers is not in the document]
Comment (keyboardDrummer) [reply Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs:103]: I see. Then I think the `SendThisVC("(reset)");` call is not necessary. But also I wonder, could we keep `SendThisVC("(reset)");` and instead remove `SetupProcess()`? Why would we want to create a new solver process for each batch?

[the remark this answers is not in the document]
Comment (atomb) [reply Source/Provers/SMTLib/SMTLibBatchTheoremProver.cs:103]: The motivation for re-starting the solver is to support solvers that can't solve more than one query per process.

**Pull request merged**